### ISSUE TITLE
Cluster Processing Hacks

### ISSUE THREAD

**Issue by StephanPreibisch** [bigdataviewer/SPIM_Registration#40]
Hi @tpietzsch & @tomancak,

did we agree that we remove the cluster saving of XML's (add name tags to the XML filename) and instead - if cluster processing is turned on - always read the XML first and then write it again when a write is requested?

It takes out flexibility but makes it easier to use.

Cheers,
@StephanPreibisch

**Comment by StephanPreibisch** [comment]
This might also directly concern @schmiedc @StephanJanosch @APavlopoulos

**Comment by schmiedc** [comment]
So the cluster would write into the same xml file? 
My plan is to make a proper pipeline and the pipeline would hide this step from the end user. 
But I don't care if the new solution works stable and reliable and would allow us to change the pipeline. 
What flexibility would we loose?

**Comment by StephanPreibisch** [comment]
That all nodes who want to write the XML must be able to see the same instance of the file while it is processing. So, if there is no shared filesystem it would not work.

**Comment by tpietzsch** [comment]
@StephanPreibisch Hmm... I guess this would work nicely for the registration step where the ViewRegistrations are independently updated. However, it might not work so nice in other cases.
For example, consider adding fused/deconvolved images to an existing dataset. Here, modification to the XML file is adding a new ViewSetup and adding a ViewRegistration for each fused/deconvolved image. If every job that writes a TIFF also reads/writes the XML, how would we be sure that every job does not append another ViewSetup? In this case, I think it is more appropriate to have one designated job that modifies the XML file. All other jobs only write images.
This is not really an argument against always-read-before-writing approach. Just note, that there are situations where it is beneficial when only one job touches the XML.

In more practical terms: How do we lock the file access? Read-modify-write should be an atomic operation right?

**Comment by StephanPreibisch** [comment]
You are totally right @tpietzsch ... this is a problem. If each job writes its own XML one can still try to figure out later what actually happened and fix it. If all write one and something goes wrong, that's it I guess.

Regarding the locking, there are mechanism in Java: http://stackoverflow.com/questions/128038/how-can-i-lock-a-file-using-java-if-possible - but I have no experience with that either.

**Comment by StephanPreibisch** [comment]
P.S. I really do not like my "cluster-ext-file-hack". This will break at some point for sure. I just really would like to have a better solution than that.

Just a crazy idea maybe, but how about we instigate a mechanism that does not write the XML itself but instead appends commands to the XML what it wants to do with it (e.g. add registration, add interestpoints, ...)? These set of commands might be way easier to combine than generically merging XML's. And we would define the set of rules that exist and extend if necessary. I clearly did not think that through but just to put it on the table.

**Comment by psteinb** [comment]
concurrently writing to e.g. hdf5 files is well possible with hdf5. And 
there is plenty of technology available to do it:

https://www.hdfgroup.org/HDF5/PHDF5/ (parallel hdf5)
https://www.olcf.ornl.gov/center-projects/adios/ (very generic framework 
to do parallel i/o to hdf5 and other formats, contains java wrapper!)

just a crazy thought: why not ditch the xml and move its the contents 
into the hdf5 (use a separate hdf5 file)? because if you do that, all 
the file-is-locked-or-not rational for the xml simply goes away and you 
have access to a whole plethora of libraries doing the parallel i/o for you!

my € 0.02
